hawc/apps/animal/managers.py

```diff
@@ -297,6 +284,98 @@ def migrate_terms(self, assessment: Assessment) -> pd.DataFrame:
+        if len(invalid_endpoint_ids) > 0:
+            invalid_endpoint_ids_str = ", ".join(str(_) for _ in invalid_endpoint_ids)
+            raise ValidationError(f"Invalid endpoint id(s) {invalid_endpoint_ids_str}")
+        # endpoints must be in the same assessment
+        excluded_endpoints = endpoints.exclude(assessment_id=assessment.id)
+        if excluded_endpoints.exists():
+            excluded_endpoints_str = ", ".join(str(_.pk) for _ in excluded_endpoints)
+            raise ValidationError(
+                f"Endpoints must be from the same assessment; endpoint(s) {excluded_endpoints_str} not from assessment {assessment.id}"
+            )
+        # term ids must be valid
+        term_ids = {obj["name_term_id"] for obj in objs}
+        terms = Term.filter(pk__in=term_ids)
+        valid_term_ids = terms.values_list("id", flat=True)
+        invalid_term_ids = term_ids - set(valid_term_ids)
+        if len(invalid_term_ids) > 0:
+            invalid_term_ids_str = ", ".join(str(_) for _ in invalid_term_ids)
+            raise ValidationError(f"Invalid term id(s) {invalid_term_ids_str}")
+        # terms must be the correct type
+        excluded_terms = terms.exclude(type=VocabularyTermType.name.value)
+        if excluded_terms.exists():
+            excluded_terms_str = ", ".join(str(_.pk) for _ in excluded_terms)
+            raise ValidationError(f"Term(s) {excluded_terms_str} are not type 'name'")
+
+    def update_terms(self, objs: List[Dict], assessment: Assessment) -> List:
+        # validate the endpoints and terms
+        self._validate_update_name_terms(objs, assessment)
+
+        # mapping of endpoint id to name_term_id
+        endpoint_id_to_term_id = {obj["id"]: obj["name_term_id"] for obj in objs}
+
+        # mapping of term id to parent
+        name_term_ids = {obj["name_term_id"] for obj in objs}
+        name_terms = Term.filter(pk__in=name_term_ids)
+        term_id_to_parent_id = {}
+
+        def _add_to_dict(term):
+            if term.id in term_id_to_parent_id:
+                return
+            term_id_to_parent_id[term.id] = term.parent_id
+            _add_to_dict(term.parent)
+
+        for term in name_terms:
+            _add_to_dict(term)
+
+        all_term_ids = set(term_id_to_parent_id.keys()) + set(term_id_to_parent_id.values())
+        all_terms = Term.filter(pk__in=all_term_ids)
+        term_id_to_type_and_name = {term.id: (term.type, term.name) for term in all_terms}
+
+        # set endpoint terms
+        endpoint_ids = [obj["id"] for obj in objs]
+        endpoints = self.get_queryset().filter(pk__in=endpoint_ids)
+        type_to_text_field = VocabularyTermType.value_to_text_field()
+        type_to_term_field = VocabularyTermType.value_to_term_field()
+        updated_endpoints = []
+        updated_fields = list(type_to_text_field.values()) + list(type_to_term_field.values())
+        for endpoint in endpoints:
```

Comment:
With the term_id, we can set all the other EHV fields too

The `term. ehv_endpoint_name` method returns system, organ effect, etc for these which can also be updated.  It's probably more efficient to in terms of db queries to use the `Term.ehv_dataframe()` method, but you get the idea.

Reply:
I believe my implementation is pretty efficient, since I only access terms that I know will be assigned instead of all terms in the EHV. Also in my recursive function I don't retread the same terms if two name terms share part of their tree, so I think its more efficient in that way too. If you think it would be cleaner or read better to use these methods though I can go back and change my implementation to use them.

Reply:
The cache you're using looks good, but I think the issue is more in the database requests. If you update 10 different endpoint ids, then it requires 40 independent database queries since there are 4 parents. Building the dataframe is one (big) hit as opposed to 50 little ones.

Reply:
Per our talk I swapped in the `ehv_dataframe`, I think it makes the code look cleaner too!